File: src/main/java/io/github/bot/terminal/application/domains/common/action/factory/ActionsFactory.java
package io.github.bot.terminal.application.domains.common.action.factory;


import io.github.bot.terminal.application.domains.common.action.entity.*;
import io.github.bot.terminal.application.domains.common.action.values.ActionType;
import io.github.bot.terminal.application.domains.common.services.ExcelManagerService;
import io.github.bot.terminal.application.domains.db_connection.entity.DbConnection;
import io.github.bot.terminal.application.domains.db_connection.factory.DbConnectionsFactory;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;

@Service
@RequiredArgsConstructor
public class ActionsFactory {

    private final DbConnectionsFactory dbConnectionsFactory;
    private final ExcelManagerService excelManagerService;

    public Action build(ActionDetails actionDetails) {
        if (ActionType.SQL_SELECT_AS_TEXT.equals(actionDetails.getType())) {
            SqlSelectAsTextActionDetails details
                    = (SqlSelectAsTextActionDetails) actionDetails;
            DbConnection<?> dbConnection = dbConnectionsFactory.byId(details.getDbConnectionId());
            return new SqlSelectAsTextAction(details.getSelect(), details.getResultTemplate(), dbConnection);
        }
        if (ActionType.SQL_SELECT_IN_EXCEL_FILE.equals(actionDetails.getType())) {
            SqlSelectInExcelFileActionDetails details
                    = (SqlSelectInExcelFileActionDetails) actionDetails;
            DbConnection<?> dbConnection = dbConnectionsFactory.byId(details.getDbConnectionId());
            return new SqlSelectInExcelFileAction(details.getSelect(), details.getFileNameTemplate(),
                    dbConnection, excelManagerService);
        }
        if (ActionType.SQL_SELECT_AS_ONE_NUMBER.equals(actionDetails.getType())) {
            SqlSelectAsOneNumberValueActionDetails details
                    = (SqlSelectAsOneNumberValueActionDetails) actionDetails;
            DbConnection<?> dbConnection = dbConnectionsFactory.byId(details.getDbConnectionId());
            return new SqlSelectAsOneNumberValueAction(details.getSelect(), dbConnection);
        }
        throw new IllegalArgumentException("Action type not found");
    }


}

File: src/main/kotlin/io/github/bot/terminal/application/domains/common/action/factory/ActionsFactory.kt
package io.github.bot.terminal.application.domains.common.action.factory

import io.github.bot.terminal.application.domains.common.action.entity.*
import io.github.bot.terminal.application.domains.common.action.values.ActionType
import io.github.bot.terminal.application.domains.common.services.ExcelManagerService
import io.github.bot.terminal.application.domains.db_connection.factory.DbConnectionsFactory
import org.springframework.jdbc.core.JdbcTemplate
import org.springframework.stereotype.Service

@Service
class ActionsFactory(
        private val dbConnectionsFactory: DbConnectionsFactory,
        private val excelManagerService: ExcelManagerService,
        private val h2JdbcTemplate: JdbcTemplate
) {

    fun build(actionDetails: ActionDetails): Action =
            when (actionDetails.type) {
                ActionType.SQL_SELECT_AS_TEXT -> {
                    val details = actionDetails as SqlSelectAsTextActionDetails
                    val dbConnection = dbConnectionsFactory.byId(details.dbConnectionId)
                    SqlSelectAsTextAction(
                            select = details.select,
                            resultTemplate = details.resultTemplate,
                            dbConnection = dbConnection)
                }
                ActionType.SQL_SELECT_IN_EXCEL_FILE -> {
                    val details = actionDetails as SqlSelectInExcelFileActionDetails
                    val dbConnection = dbConnectionsFactory.byId(details.dbConnectionId)
                    SqlSelectInExcelFileAction(
                            select = details.select,
                            fileNameTemplate = details.fileNameTemplate,
                            dbConnection = dbConnection,
                            excelManagerService = excelManagerService
                    )
                }
                ActionType.SQL_SELECT_AS_ONE_NUMBER -> {
                    val details = actionDetails as SqlSelectAsOneNumberValueActionDetails
                    val dbConnection = dbConnectionsFactory.byId(details.dbConnectionId)
                    SqlSelectAsOneNumberValueAction(
                            select = details.select,
                            dbConnection = dbConnection
                    )
                }
                ActionType.JOIN_SQL_SELECTS_AS_TEXT -> {
                    val details = actionDetails as JoinSqlSelectAsTextActionDetails
                    JoinSqlSelectAsTextAction(
                            queries = details.queries.map {
                                SqlSelect(
                                      name = it.name,
                                      select = it.select,
                                      dbConnection = dbConnectionsFactory.byId(it.dbConnectionId)
                                )
                            },
                            select = details.select,
                            resultTemplate = details.resultTemplate,
                            h2InMemory = H2InMemory(h2JdbcTemplate)
                    )
                }
            }
}